community/kernel/src/main/java/org/neo4j/kernel/impl/store/NeoStore.java

```diff
@@ -900,12 +900,32 @@ public long nextCommittingTransactionId()
     @Override
     public void transactionCommitted( long transactionId, long checksum )
     {
-        if ( lastCommittedTx.offer( transactionId, checksum ) )
+        checkInitialized( lastCommittingTxField.get() );
+        HighestCommittedTransaction current = highestCommittedTransaction.get();
+        if ( current.transactionId() < transactionId )
         {
-            long[] transactionData = lastCommittedTx.get();
-            setRecord( Position.LAST_TRANSACTION_ID, transactionData[0] );
-            setRecord( Position.LAST_TRANSACTION_CHECKSUM, transactionData[1] );
-            lastTransactionChecksum = checksum;
+            HighestCommittedTransaction update = new HighestCommittedTransaction( transactionId, checksum );
+            while ( !highestCommittedTransaction.compareAndSet( current, update ) )
+            {
+                current = highestCommittedTransaction.get();
+                // let's check if our update is still needed
+                if ( current.transactionId() >= transactionId )
+                {
+                    return;
+                }
+            }
+
+            /*
+             * We need to synchronize here in order to guarantee that the two field are written consistently
+             * together. Note that having the exclusive lock on tha page is not enough for 2 reasons:
+             * 1. the records might be in different pages
+             * 2. some other thread might kick in while we have been written only one record
```

Comment:
This would be a problem if (1) is true, right? And since I argue that it's not, then this isn't true either. I.e. we could solve this be solely relying on the exclusive page lock here for updating these two records. But it's not super critical, I think